governator-core/src/main/java/com/netflix/governator/internal/PostConstructLifecycleActions.java

```diff
@@ -1,57 +1,119 @@
+import java.util.HashSet;
+import java.util.LinkedList;
 import java.util.List;
+import java.util.Set;
 
 import javax.annotation.PostConstruct;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Supplier;
 import com.netflix.governator.LifecycleAction;
+import com.netflix.governator.LifecycleFeature;
+import com.netflix.governator.internal.TypeInspector.TypeVisitor;
 
 /**
- * Special AbstractLifecycleFeature to support @PostConstruct annotation processing.
- * Note that this feature is implicit in LifecycleModule and therefore does not need 
- * to be added using the LifecycleFeature multibinding.
+ * Special LifecycleFeature to support @PostConstruct annotation processing.
+ * Note that this feature is implicit in LifecycleModule and therefore does not
+ * need to be added using the LifecycleFeature multibinding.
  * 
  * @author elandau
  */
-public final class PostConstructLifecycleActions extends AbstractLifecycleFeature {
-
+public final class PostConstructLifecycleActions implements LifecycleFeature {
+    private static final Logger LOG = LoggerFactory.getLogger(PostConstructLifecycleActions.class);
     public static PostConstructLifecycleActions INSTANCE = new PostConstructLifecycleActions();
-    
+
     private PostConstructLifecycleActions() {
-        
     }
-    
+
     @Override
-    protected List<LifecycleAction> getMethodActions(final Class<?> type, final Method method) {
-        if (null != method.getAnnotation(PostConstruct.class)) {
-            method.setAccessible(true);
-            return Collections.<LifecycleAction>singletonList(new LifecycleAction() {
-                @Override
-                public void call(Object obj) throws IllegalAccessException, IllegalArgumentException, InvocationTargetException {
-                    method.invoke(obj);
-                }
-                
-                @Override
-                public String toString() {
-                    return new StringBuilder()
-                        .append("PostConstruct[")
-                        .append(type.getName())
-                        .append("#")
-                        .append(method.getName())
-                        .append("]")
-                        .toString();
-                }
-            });
-        }
-        return Collections.emptyList();
+    public List<LifecycleAction> getActionsForType(final Class<?> type) {
+        return TypeInspector.accept(type, new PostConstructVisitor());
     }
-    
+
     @Override
     public String toString() {
         return "PostConstruct";
     }
 
+    private static class PostConstructVisitor implements TypeVisitor, Supplier<List<LifecycleAction>> {
+        private Set<String> visitContext = new HashSet<>();
+        private LinkedList<LifecycleAction> typeActions = new LinkedList<>();
+
+        @Override
+        public boolean visit(final Class<?> clazz) {
+            return !clazz.isInterface();
```

Comment:
'extends' is supported but 'implements' is not; it's pretty strongly worded
in jsr-250:

*The interfaces implemented by a class never contribute annotations to the
class itself or any of its members.*

we should likely go a step further and enforce that a class only specifies
a single @PostConstruct, but the implementation is a bit more involved.
I'm deferring that enforcement for now.

On Wed, Jun 1, 2016 at 5:49 PM elandau <notifications@github.com> wrote:

> In
> governator-core/src/main/java/com/netflix/governator/internal/PostConstructLifecycleActions.java
> <https://github.com/Netflix/governator/pull/297#discussion_r65468722>:
>
> >      @Override
> >      public String toString() {
> >          return "PostConstruct";
> >      }
> >
> > +    private static class PostConstructVisitor implements TypeVisitor, Supplier<List<LifecycleAction>
>
> > {
> > +        private Set<String> visitContext = new HashSet<>();
> > +        private LinkedList<LifecycleAction> typeActions = new LinkedList<>();
> > +
> > +        @Override
> > +        public boolean visit(final Class<?> clazz) {
> > +            return !clazz.isInterface();
>
> Do we not want to support 'extends'?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Netflix/governator/pull/297/files/831626d4e6cfd260e5d9b5b8516286c977ba2e89#r65468722>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACIO-2pIcifkrfDUelkSiKdSryMdhxyEks5qHiiZgaJpZM4InDUB>
> .
>